File: Linked List.cpp
#include <iostream>
#include <string>
using namespace std;

// Short exception class for the Linked List
class LinkedListException {
private:
	string message;

public:
	string what(){
		return message;
	}

	LinkedListException(string m) {
		message = m;
	}
};

// The student class will contain the information we want the nodes to represent
class Student {
private:
	string firstName;
	string lastName;

public:
	// Set method for first name
	void setFirstName(string f) {
		firstName = f;
	}

	// Set method for last name
	void setLastName(string l) {
		lastName = l;
	}

	// Get method for first name
	string getFirstName() const {
		return firstName;
	}

	// Get method for last name
	string getLastName() const {
		return lastName;
	}


	// Constructor which initializes the student's first and last name based on
	// the input parameters
	Student(string f, string l) {
		firstName = f;
		lastName = l;
	}

	// Constructor which initializes the student's first and last name as an empty string
	Student() {
		firstName = "";
		lastName = "";
	}
};

// The NodeType class will represent the nodes in the linked list
// Each node will have a Student member as the data as well as a 
// pointer to the next node in the list.
// We create the NodeType objects dynamically.
class NodeType {
public:
	// Empty constructor declaration of Student member "data"
	Student data;
	NodeType* next;
	NodeType() {
		next = nullptr;
	}
};

// The LinkedList class will contain the members and methods for making a Linked List
class LinkedListType {
private:
	int size;
	NodeType* first;
	NodeType* last;

public:

	// Checks if the list is empty by seeing
	// if the first and last nodes are null
	bool isEmpty() const {
		return (first == nullptr && last == nullptr);
	}

	// Get method for the list size
	int getSize() const {
		return size;
	}

	// Allows us to traverse the linked list and print out the data
	void display() const {
		NodeType* current = first;
		while (current != nullptr) {
			cout << current->data.getFirstName() << " " << current->data.getLastName() << endl;
			current = current->next;
		}
	}

	// The addLast function will create a new node dynamically, and then will
	// assign the new node's data to be the student object that we declare. Then, it 
	// will insert the node into the list at the last position.
	void addLast(const Student studentData) {
		NodeType* newNode = new NodeType();
		newNode->data = studentData;

		// If the list is empty, have the first and last pointers point to
		// newNode.
		if (isEmpty()) {
			first = newNode;
			last = newNode;
		}

		// If list is NOT empty, have the node pointed to by "last" change its next link
		// so that it points to newNode. Then, change "last" to point to the current final 
		// node in the list which we just added.
		else {
			last->next = newNode;
			last = newNode;
		}

		// Increase size by 1.
		size++;

	}

	// The addFirst function will create a new node dynamically, and then we will assign
	// the new node's data to be the student object that we declare. Then, it will
	// insert the node into the first position of the linked list.
	void addFirst(const Student studentData) {
		NodeType* newNode = new NodeType();
		newNode->data = studentData;

		// If the list is empty, have the first and last pointers point to
		// newNode.
		if (isEmpty()) {
			first = newNode;
			last = newNode;
		}

		// If the list is not empty, have new node's link point to first and then
		// update first pointer to point to newNode which we just added.
		else {
			newNode->next = first;
			first = newNode;
		}

		// Increase size by 1.
		size++;
	}
	
	// This function returns a copy of the first element in the list. If the list is empty,
	// it will throw an exception.
	Student getFirst() const {

		// If the function is empty, throw an exception
		if (isEmpty())
			throw LinkedListException("The list is empty.");

		// Return data of first node if list is not empty.
		else
			return first->data;

	}

	// This function returns a copy of the last element in the list. If the list is empty,
	// it will throw an exception.
	Student getLast() const {

		// If the function is empty, throw an exception
		if (isEmpty())
			throw LinkedListException("The list is empty.");

		// Return data of last node if list is not empty.
		else
			return last->data;

	}

	// DeleteFirst function
	void DeleteFirst() {
		// Throw an exception if list is empty
		// Since we cannot remove non-existing nodes
		if (isEmpty())
			throw LinkedListException("The list is empty.");

		// If size is only 1 node, delete it and set first and last
		// to null pointers since there is nothing to point to in the list
		// and decrease size by 1
		else if (size == 1) {
			delete first;
			first = nullptr;
			last = nullptr;
			size--;
		}

		// If size > 1, We must use a temp variable to point to "first" node and delete
		// the temp once we have first point to first->next
		// Reason: If we simply delete first, then we will have no method to access
		// The next node in the list since the firts node's data, including its link
		// Will also be deleted, and we won't be able to set "first" to first->next
		// And simply having first = first->next will not delete first node.
		else {
			NodeType* temp = first;
			first = first->next;
			delete temp;
			size--;
		}
	}

	// DeleteLast function
	void DeleteLast() {
		// Throw an exception if list is empty
		// Since we cannot remove non-existing nodes
		if (isEmpty()) 
			throw LinkedListException("The list is empty.");

		// If size is only 1 node, delete it and set first and last
		// to null pointers since there is nothing to point to in the list
		// and decrease size by 1
		else if (size == 1) {
			delete first;
			first = nullptr;
			last = nullptr;
			size--;
		}

		// If size > 1, keep track of previous pointer pointed to by last
		// Since we don't want the previous node to be pointing to a dangling 
		// pointer after we delete node pointed to by "last"
		else {
			NodeType* previous = first;

			// Keep traversing through linked list until the next link
			// Points to "last"
			while (previous->next != last)
				previous = previous->next;

			// Delete node pointed to by "last" once previous->next points
			// to last, then have "last" point to previous since last is
			// now a dangling pointer.
			delete last;
			last = previous;

			// We must then set previous->next to nullptr to
			// avoid it pointing to a dangling pointer
			previous->next = nullptr;
			size--;
		}

	}

	// Constructor for Linked List class
	LinkedListType() {
		size = 0;
		first = nullptr;
		last = nullptr;
	}

};

int main()
{
	cout << "Linked List Application!\n";
	cout << "===========================\n";

	// Create three student objects
	Student student_1("Patrick", "Ducusin");
	Student student_2("Bruce", "Banner");
	Student student_3("Tony", "Stark");

	// Declare a Queue, represented by a linked list 
	LinkedListType queue;

	// Add the three student objects into linked list using addFirst() function
	// Since we want to represent the data in a Queue
	queue.addFirst(student_1);
	queue.addFirst(student_2);
	queue.addFirst(student_3);

	// Display contents of front element in queue using getLast() function
	// Since the first element becomes the "final" one in the queue, and should
	// be the first one to get retrieved again
	cout << queue.getLast().getFirstName() << " " << queue.getLast().getLastName() << endl;

	// Delete front element using the deleteLast() function
	queue.DeleteLast();

	// Repeat process with second element
	cout << queue.getLast().getFirstName() << " " << queue.getLast().getLastName() << endl;
	queue.DeleteLast();

	// Repeat process with third element
	cout << queue.getLast().getFirstName() << " " << queue.getLast().getLastName() << endl;
	queue.DeleteLast();

	// Show that queue is empty by trying to retrieve information
	try {
		cout << queue.getLast().getFirstName() << " " << queue.getLast().getLastName() << endl;
	}
	catch (LinkedListException obj) {
		cout << obj.what();
	}
}

File: linked_list.py
class Node:
    # The constructor of the data node needs
    # to initialize the data that the node stores, but
    # we will keep the link pointing to None as a default
    # parameter
    def __init__(self, userData, nextNode=None):
        self.data = userData
        self.next = nextNode

    # Overridden method to print out a node
    def __repr__(self):
        return "[%s]" % self.data

class LinkedList:

    # Constructor of Linked list class
    def __init__(self, data: list = None):
        self.head = None
        self.tail = None

        # If the data parameter is not empty,
        # then we will make sure to store the data
        # as nodes in our linked list. We will add it to
        # the end.
        if data:
            for i in range(len(data)):
                node = Node(data[i])
                self.addLast(node)

    # The object instance printing reference
    def __repr__(self):
        node = self.head
        LList = []
        while node is not None:
            if node is self.head:
                LList.append("[Head: %s]" % node.data)
            elif node is self.tail:
                LList.append("[Tail: %s]" % node.data)
            else:
                LList.append("[%s]" % node.data)
            node = node.next

        LList.append("None")
        return " -> ".join(LList)

    # To check if the Linked List is empty
    def isEmpty(self):
        return self.head is None

    # Function to insert data nodes at the beginning of the Linked List
    def addFirst(self, node):
        if self.isEmpty():
            self.head = node
            self.tail = node

        # We need to have the node point to whatever the current head is at
        # and then assign head to point to the node. This ensures the node is
        # now at the beginning of the list.
        else:
            node.next = self.head
            self.head = node


    # Function to insert data Nodes at the end of the Linked List
    def addLast(self, node):
        if self.isEmpty():
            self.head = node
            self.tail = node

        # To insert at the end, we need to have the tail node
        # point to the input node. Then, we will have tail point to
        # the node in order to show that it is the final node.
        else:
            self.tail.next = node
            self.tail = node


    # Function to delete data nodes starting at the beginning of the list
    def deleteFirst(self):
        if self.isEmpty():
            raise Exception("List is empty!")
        elif self.head == self.tail:
            self.head = None
            self.tail = None
        # We set the head to point to whatever it is linked to in order to indicate that the previous
        # head is no longer apart of the list. This works because we the previous head node will
        # now be inaccessible.
        else:
            self.head = self.head.next


    # Function to delete data nodes at the end of the list
    def deleteLast(self):
        if self.isEmpty():
            raise Exception("List is empty!")
        elif self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            # We will have a prev pointer starting at the head
            # and a current_node starting at the node pointed by
            # head.next
            # IFF the current_node.next is not None, we will update
            # prev to point to current_node, then have current_node
            # point to whatever node it is pointing to next.
            # IFF current_node.next is None, we will set prev.next to
            # be none, as to show we are deleting current (since prev.next
            # is pointing to current_node), then we will have tail point
            # to prev
            prev = self.head
            current_node = self.head.next
            while current_node.next is not None:
                prev = current_node
                current_node = current_node.next
            prev.next = None
            self.tail = prev

    # Function that is to determine the size of the linked list
    def size(self):
        # if list is empty, return 0
        if self.isEmpty():
            return 0
        else:
            # If not empty, declare a size variable = 0
            # Have a current pointer and initialize it to the
            # head node of the list
            size = 0
            current = self.head

            # While current is pointing to a Node, we will
            # first increase the size to indicate that we
            # have counted a valid node. Then, have current
            # point to the next node.
            while current is not None:
                size += 1
                current = current.next

            return size

    # This function will return a node at our specified index
    def node_at_index(self, index):
        # If our list is empty, raise an exception
        if self.isEmpty():
            print("List is empty!")
            return None

        # If index is 0, return the head
        elif index == 0:
            return self.head

        else:
            # If index is greater than 0, have a current
            # pointer and initialize it to the head of the list
            # Then have a position variable set to 0
            current = self.head
            position = 0

            # While position is less than the target index, we will:
            # FIRST, have current point to the next node in the linked list.
            # SECOND, we will increment the position by 1 to indicate the current
            # node's new index in the linked list (we incremented by 1).
            while position < index:
                current = current.next
                position += 1

            # Once the while loop terminates, we are at the node at the
            # target index. Return current
            return current

    # This function will insert a node in the middle of the
    # linked list
    def insertInMid(self, node):
        # First, check if our linked list is EMPTY! If it is,
        # we will have the head and tail both point to the new
        # node in the linked list
        if self.isEmpty():
            self.head = node
            self.tail = node

        # If our linked list is NOT empty, then we need to determine
        # the middle index by doing integer division. We will call
        # the size method and divide the result by 2, and then subtract by
        # 1 to get the middle index since indices are 0 based.
        else:
            mid = (self.size() // 2) - 1

            # Have our input node's next index point to whatever the
            # middle node's next pointer was originally pointing to
            # (Since it will be in between mid and mid.next).
            # Finally, have the middle node's next pointer now
            # point to our input node. This effectively turns our
            # input node to the new middle node in the linked list.
            node.next = self.node_at_index(mid).next
            self.node_at_index(mid).next = node


myLinkedList = LinkedList([1, 2, 3, 4])
myLinkedList.addLast(Node("A"))
myLinkedList.addLast(Node("B"))
myLinkedList.addLast(Node("C"))
myLinkedList.addLast(Node("D"))

print("Node at index 7 = " + str(myLinkedList.node_at_index(7)))
print("Insert at Middle Index: Cool Beans")
myLinkedList.insertInMid(Node("Cool Beans"))

print(myLinkedList)
print("Count = " + str(myLinkedList.size()))

myLinkedList.deleteLast()

print(myLinkedList)
print("Count = " + str(myLinkedList.size()))

myLinkedList.deleteLast()

print(myLinkedList)

myLinkedList.deleteLast()

print(myLinkedList)

myLinkedList.deleteLast()

print(myLinkedList)




